### ISSUE TITLE
newly updated crates are not built if there are too many of them

### ISSUE THREAD

**Issue by Byron** [onur/docs.rs#89]
I have recently published a great amount of crates in an automated fashion, and noticed that docs-rs didn't really catch up with them.

<img width="278" alt="screen shot 2016-12-24 at 22 16 42" src="https://cloud.githubusercontent.com/assets/63622/21468795/837b8074-ca27-11e6-9f8d-0036e2de9104.png">

An example would be the [google-logging2](https://crates.io/crates/google-logging2/1.0.1+20161206) crate, which exists in version _1.0.1_, whereas docs.rs still shows it [as version 1.0.0](https://docs.rs/google-logging2).

A glimpse into the code indicates that docs.rs detects newly updated crates by looking at the `.just_updated` field of the [crates.io summary endpoint](https://crates.io/summary) ([see code](https://github.com/onur/docs.rs/blob/master/src/docbuilder/queue.rs#L32)), which appears to be a problem if there are many new releases in an interval shorter than the one used to query the endpoint.

As I just now moved all my links to point to docs.rs, it would be great if we could find out what could be done to alleviate the issue :).

Thank you.

**Comment by Byron** [comment]
I am actually working on a PR to fix this. The idea is to clone the crates index and keep track of the last commit we have seen.
Each time we try to get information about changed crates, we will fetch latest and diff the changes. That way we will never be able to miss anything, as we are then based on the only source of truth.

I will also setup a WIP PR in a moment so you can stay informed on the progress.

**Comment by onur** [comment]
Thanks for reporting.

This is similar to #84. Since docs.rs started getting new crate list from crates.io/summary this issue also occurred. docs.rs is checking new crates every 5 minutes and crates.io/summary is only returning last updated 10 crates. If crates.io gets more than 10 uploads in 5 minutes, docs.rs is skipping them.

I will revert this changes and start getting new crates from crates.io-index repository again.

Leaving this commit id for reference: 0dfb64c.

**Comment by Byron** [comment]
Thanks for elaborating on this!
Does it mean you don't need me to work on this? Actually I would love to get involved, but of course only if there is no code to do this lying around already.

**Comment by Byron** [comment]
Ah, I see there is code for that already. However, I wouldn't mind giving it a try too. Can you give me an advise on how you would want me to proceed?

**Comment by onur** [comment]
Thanks for offering help! I changed old code because it wasn't working properly. You can see the old code from here: [queue.rs](https://github.com/onur/docs.rs/blob/0dfb64c64667f5b79361bdbaae26a80546c50eb6/src/docbuilder/queue.rs).

I have been trying to parse [DiffLine](https://docs.rs/git2/0.6.3/git2/struct.DiffLine.html) with [Diff::print](https://docs.rs/git2/0.6.3/git2/struct.Diff.html#method.print) but this seems really wrong.

Maybe you can figure out how to get new crates properly by using git2 crate and crates.io-index. There must be a better way or approach to do this but I have no idea.

**Comment by Byron** [comment]
Alright :)! I happen to have been working with git internals quite a bit in the past, and I always wanted to work with it once again from Rust.
I have already seen the previous issues and code and will work on a new implementation, keeping a close eye on leaked file handles.
Shall keep you posted!

**Comment by Byron** [comment]
Here is the first version of the [crates-index-diff](https://github.com/Byron/crates-index-diff-rs) crate. It's already providing all the functionality you would need.
Tomorrow I will push it towards 1.0 (refactoring, cleanup, docs), and prepare PR for docs.rs to use it.
Please feel free to comment on it already.

**Comment by onur** [comment]
@Byron this seems really promising, thank you very much.

There is only one issue. Docs.rs dependencies (cargo, hyper and iron especially) are using openssl 0.7 and latest git2 crate started using openssl 0.9. Currently we can't use both of them together. We have to use git2 0.5 until hyper starts using openssl 0.9. There is an ongoing PR for hyper: hyperium/hyper#975. But updating iron to 0.4 in docs.rs will require some work.

So basically your crate must depend on git2 0.5 for now, we can update it later when all dependencies supports openssl 0.9.

**Comment by Byron** [comment]
[Version 1.0](https://github.com/Byron/crates-index-diff-rs) of the crate has now been released. It's dependency to git was already adjusted to match the requirements of docs.rs.

Now I am working on a [quick CLI](https://github.com/Byron/crates-io-cli-rs) to show how it is used, and to have it ready in case you would like to keep the git-related accesses out of process (and also because I want it :)).

**Comment by Byron** [comment]
Alright, the [crates-io-cli](https://github.com/Byron/crates-io-cli-rs) is ready, and allows to test manually use the functionality of the `crates-index-diff` crate.

Next up is a PR to actually use the new functionality in `docs.rs`.
